python/cudf/cudf/dataframe/datetime.py

```diff
@@ -97,32 +122,25 @@ def normalize_binop_value(self, other):
             other = np.datetime64(other)
 
         if isinstance(other, pd.Timestamp):
-            ary = utils.scalar_broadcast_to(
-                other.value * self._pandas_conversion_factor,
-                shape=len(self),
-                dtype=self._npdatetime64_dtype,
-            )
+            m = _numpy_to_pandas_conversion[self.time_unit]
+            ary = utils.scalar_broadcast_to(other.value * m,
+                                            shape=len(self),
+                                            dtype=self.dtype)
         elif isinstance(other, np.datetime64):
-            other = other.astype(self._npdatetime64_dtype)
-            ary = utils.scalar_broadcast_to(
-                other, shape=len(self), dtype=self._npdatetime64_dtype
-            )
+            other = other.astype(self.dtype)
+            ary = utils.scalar_broadcast_to(other,
+                                            shape=len(self),
+                                            dtype=self.dtype)
         else:
             raise TypeError("cannot broadcast {}".format(type(other)))
 
-        buf = Buffer(ary)
-        result = self.replace(data=buf, dtype=self.dtype)
-        return result
+        return self.replace(data=Buffer(ary), dtype=self.dtype)
 
     @property
     def as_numerical(self):
-        from cudf.dataframe import numerical
-
-        return self.view(
-            numerical.NumericalColumn,
-            dtype="int64",
-            data=self.data.astype("int64"),
-        )
+        from cudf.dataframe.numerical import NumericalColumn
+        data = self.data.astype(utils.datetime_to_numerical_dtype(self.dtype))
```

Comment:
This property shouldn't require a typecast, this should just view the current underlying data as an int64 instead of the datetime type underneath it.

Reply:
This is required because the `NumericalColumn` constructor asserts `self._dtype == data.dtype`, and IIRC `self.data` here has the datetime64 dtype of the DatetimeColumn.

But since `Buffer.astype` calls `cudautils.astype`, which returns a view when casting `np.datetime64` -> `np.int64`, then barring anyone messing about with column internals, this should be guaranteed zero-copy.